### PR TITLE
feat: Add slippage to maverick fetcher

### PR DESCRIPTION
**Pull request by koonopek**

### CODE REVIEW

File: packages/oracle-node/src/fetchers/dex-on-chain/DexOnChainFetcher.ts [hunk 1]
```diff
@@ -48,10 +48,10 @@ export abstract class DexOnChainFetcher<T> extends MultiRequestFetcher {
     } catch (e) {}
 
     try {
-      const liquidity = this.calculateLiquidity(dataFeedId, response);
-      if (liquidity) {
-        metadata.liquidity = liquidity.toString();
-      }
+      metadata.liquidity = this.calculateLiquidity(
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/dex-on-chain/DexOnChainFetcher.ts:51]: Why this change?

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts [hunk 2]
```diff
@@ -78,3 +166,111 @@ export class MaverickFetcher extends DexOnChainFetcher<MaverickResponse> {
...
+  }
+
+  buildRequest(): RedstoneCommon.CallParams[] {
+    const requests: RedstoneCommon.CallParams[] = [
+      { function: "getSqrtPrice", params: [this.tokenConfig.poolAddress] },
+    ];
+
+    if (this.amountIn0Token) {
+      requests.push(
+        {
```
Comment (rpiszczatowski) [inline packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:194]: Do we need to call `calculateSwap` for small values? In the end, `spotPrice` is what we need to calculate slippage.

File: packages/oracle-node/src/fetchers/dex-on-chain/DexOnChainFetcher.ts (hunk 1 above)
Comment (koonopek) [reply packages/oracle-node/src/fetchers/dex-on-chain/DexOnChainFetcher.ts:51]: less lines + just resolved conflict

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts (hunk 2 above)
Comment (koonopek) [reply packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:194]: Probably yes, however, I'm not 100% sure what sqrtPrice is doing, because code is closed. Thus I think it will be safer to relay on the same methods to calculate slippage. In the end, we only calculate the relative difference.

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts [hunk 3]
```diff
@@ -18,49 +26,78 @@ interface FetcherConfig {
...
       MAVERICK_POOL_INFORMATION_ABI,
       this.config.provider
     );
 
-    const sqrtPrice = await contract.getSqrtPrice(poolAddress);
-    const priceInPairedToken = new Decimal(sqrtPrice.toString())
-      .div(MAVERICK_PRECISION_DIVIDER)
-      .toPower(2);
+    const token0Scaler = new MathUtils.PrecisionScaler(token0Decimals);
+    const token1Scaler = new MathUtils.PrecisionScaler(token1Decimals);
```
Comment (cehali) [inline packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:60]: Why do you build scalers here only to get decimals from them? Can we just pass `token0Decimals` and `token1Decimals` to `tryConvertUsdToTokenAmount`?

File: packages/oracle-node/src/fetchers/maverick/pool-information.abi.ts [hunk 4]
```diff
@@ -1,3 +1,15 @@
...
+  /// @param amount amount of token that is either the input if exactOutput
+  //is false or the output if exactOutput is true
+  /// @param tokenAIn bool indicating whether tokenA is the input
+  /// @param exactOutput bool indicating whether the amount specified is the
+  //exact output amount (true)
+  /// @param sqrtPriceLimit limiting sqrt price of the swap.  A value of 0
+  //indicates no limit.  Limit is only engaged for exactOutput=false.  If the
+  //limit is reached only part of the input amount will be swapped and the
+  //callback will only require that amount of the swap to be paid.
+  "function calculateSwap(address pool, uint128 amount, bool tokenAIn, bool exactOutput, uint256 sqrtPriceLimit) external returns (uint256 returnAmount)",
```
Comment (cehali) [inline packages/oracle-node/src/fetchers/maverick/pool-information.abi.ts:14]: Comments in abi file probably would be easily missed, maybe we can move them to `MultiCallHandler`

File: packages/oracle-node/test/fetchers/maverick.spec.ts [hunk 5]
```diff
@@ -36,14 +41,16 @@ describe("Maverick", () => {
     await closeLocalLevelDB();
```
[the remark this answers is not in the document]
Comment (koonopek) [reply packages/oracle-node/test/fetchers/maverick.spec.ts:41]: True, done

File: packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts [hunk 6]
```diff
@@ -15,6 +15,22 @@ export const convertUsdToTokenAmount = (
...
+  amountInUsd: number
+): string | undefined => {
+  const rawPrice = getRawPrice(assetId)?.value;
+  if (!rawPrice) {
+    return undefined;
+  }
+  return new Decimal(amountInUsd)
+    .div(rawPrice)
+    .mul(decimalsMultiplier)
+    .round()
```
[the remark this answers is not in the document]
Comment (koonopek) [reply packages/oracle-node/src/fetchers/SlippageAndLiquidityCommons.ts:30]: Yep, cause we need integer as result

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts (hunk 3 above)
Comment (koonopek) [reply packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:60]: good point

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts [hunk 7]
```diff
@@ -78,3 +166,111 @@ export class MaverickFetcher extends DexOnChainFetcher<MaverickResponse> {
     return new Decimal(lastPriceFromCache.value);
   }
 }
+
+class MultiCallHandler {
```
[the remark this answers is not in the document]
Comment (koonopek) [reply packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:170]: Why? I think this is strictly bounded to other class in this file. It is not meant to be reused.

File: packages/oracle-node/src/fetchers/maverick/pool-information.abi.ts (hunk 4 above)
Comment (koonopek) [reply packages/oracle-node/src/fetchers/maverick/pool-information.abi.ts:14]: done

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts [hunk 8]
```diff
@@ -18,49 +26,75 @@ interface FetcherConfig {
...
+      amountIn0Token,
+      amountIn1Token
+    );
+
+    const response = await RedstoneCommon.multiCallOneContract(
+      contract,
+      multicallHandler.buildRequest()
+    );
+
+    return await multicallHandler.parseResponse(response);
```
Comment (cehali) [inline packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:82]: In most fetchers `makeRequest` is only used for fetching the data, not parsing it. I am wondering if it would be cleaner to parse the response and extract proper values in `calculate....` functions

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (cehali) [reply packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:167]: Putting in separate files is not always used to reuse something. It is more aesthetic manner, one file for one class. It would make this file much smaller and as a result easier to cope with

File: packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts (hunk 8 above)
Comment (koonopek) [reply packages/oracle-node/src/fetchers/maverick/MaverickFetcher.ts:82]: Then i would have to parse it more then one time, once in calculateSpotPrice once in calculateSlippage.

**Review by cehali — APPROVED** [review]
LGTM
